File: GotDotNet.com/ES Hands On Lab/VB/Solutions/Exercise 3/HelloEnterpriseServices/HelloTx.vb
Imports System.Windows.Forms
Imports System.EnterpriseServices

<Transaction(TransactionOption.Required)> _
Public Class HelloTx
    Inherits ServicedComponent

    Public Sub SayHello(ByVal Name As String)
        Dim txId As Guid
        txId = ContextUtil.TransactionId
        Select Case MessageBox.Show(String.Format( _
                    "Hello {0} - Should I commit transaction {1}?", Name, txId), _
                    "Transaction", MessageBoxButtons.YesNo)
            Case DialogResult.Yes
                ContextUtil.SetComplete()
            Case DialogResult.No
                ContextUtil.SetAbort()
        End Select
    End Sub

    Private Function RandomCommit(ByRef Message As String)
        Dim Commit As Boolean
        Dim r As New Random(DateTime.Now.Millisecond)
        ' Commit if an even number is returned from the random generator
        Commit = ((r.Next() Mod 2) = 0)
        Message = String.Format("RandomOutcome Result: {0} Transaction ID: {1}", IIf(Commit, "commit", "abort"), ContextUtil.TransactionId)
        Return Commit
    End Function

    <AutoComplete()> _
    Public Sub RandomOutcome()
        Dim Message As String
        If Not RandomCommit(Message) Then
            Throw New Exception(Message)
        Else
            Console.WriteLine(Message)
        End If
    End Sub
End Class

File: GotDotNet.com/ES Hands On Lab/CS/Solutions/Exercise 3/HelloEnterpriseServices/HelloTx.cs
using System;
using System.Windows.Forms;
using System.EnterpriseServices;
using System.Collections;
using System.Diagnostics;

namespace HelloEnterpriseServices
{
	[Transaction]
	public class HelloTx : ServicedComponent
	{
		public HelloTx()
		{
			//
			// TODO: Add constructor logic here
			//
		}

		public void SayHello(string Name)
		{
			Guid txID = ContextUtil.TransactionId;
			string Msg = string.Format("Hello {0} - Commit transaction {1}?", Name, txID);
			switch (MessageBox.Show(Msg,"Transaction", MessageBoxButtons.YesNo))
			{
				case DialogResult.Yes:
					ContextUtil.SetComplete();
					break;
				case DialogResult.No:
					ContextUtil.SetAbort();
					break;
			}
		}

		private bool RandomCommit(out string Message)
		{
			bool Commit;
			Random r = new Random(DateTime.Now.Millisecond);
			// Commit if an even number is returned from the random generator
			Commit = ((r.Next() % 2) == 0) ? true : false;
			Message = String.Format("RandomOutcome Result: {0} Transaction ID: {1}", Commit ? "commit" : "abort", ContextUtil.TransactionId);
			return Commit;
		}

		[AutoComplete]
		public void RandomOutcome()
		{
			string Message;
			if (!RandomCommit(out Message))
				throw new Exception(Message);
			else
				Console.WriteLine(Message);
		}
	}
}
